src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,257 @@
+                    // To enable empirical fine tuning, we apply a configurable multiplier (_windowScaleThresholdMultiplier) to the window size, which defaults to 1.0
+                    //
+                    // The condition to extend the window is:
+                    // (_deliveredBytes / dt) * rtt > _streamWindowSize * _windowScaleThresholdMultiplier
+                    //
+                    // Which is reordered into the form below, to avoid the division:
+                    if (_deliveredBytes * rtt.Ticks > _streamWindowSize * dt.Ticks * _windowScaleThresholdMultiplier)
+                    {
+                        int extendedWindowSize = Math.Min(_maxStreamWindowSize, _streamWindowSize * 2);
+                        windowUpdateIncrement += extendedWindowSize - _streamWindowSize;
+                        _streamWindowSize = extendedWindowSize;
+
+                        if (NetEventSource.Log.IsEnabled()) _stream.Trace($"[FlowControl] Updated Stream Window. StreamWindowSize: {StreamWindowSize}, StreamWindowThreshold: {StreamWindowThreshold}");
+
+                        Debug.Assert(_streamWindowSize <= _maxStreamWindowSize);
+                        if (_streamWindowSize == _maxStreamWindowSize)
+                        {
+                            if (NetEventSource.Log.IsEnabled()) _stream.Trace($"[FlowControl] StreamWindowSize reached the configured maximum of {_maxStreamWindowSize}.");
+
+                            // We are at maximum configured window, stop further scaling:
+                            _windowScalingEnabled = false;
+                            _connection._rttEstimator.MaximumWindowReached();
+                        }
+                    }
+                }
+
+                _deliveredBytes = 0;
+
+                Task sendWindowUpdateTask = _connection.SendWindowUpdateAsync(_stream.StreamId, windowUpdateIncrement);
+                _connection.LogExceptions(sendWindowUpdateTask);
+
+                _lastWindowUpdate = currentTime;
+            }
+
+            private static TimeSpan StopwatchTicksToTimeSpan(long stopwatchTicks)
+            {
+                long ticks = (long)(StopWatchToTimesSpan * stopwatchTicks);
+                return new TimeSpan(ticks);
+            }
+        }
+
+        private struct RttEstimator
+        {
+            private enum State
+            {
+                Disabled,
+                Init,
+                Waiting,
+                PingSent
+            }
+
+            private const double PingIntervalInSeconds = 1;
+            private static readonly long PingIntervalInTicks =(long)(PingIntervalInSeconds * Stopwatch.Frequency);
+
+            private Http2Connection _connection;
+
+            private State _state;
+            private long _pingSentTimestamp;
+            private long _pingCounter;
+            private int _initialBurst;
+            private volatile bool _terminating;
+            private long _minRtt;
```

Comment:
Given that this is used as a TimeSpan, why not store it as a TimeSpan too?

Reply:
The update of the value should be thread safe (atomic), so I had to change it to `long` so we can use it with `Interlocked.Exchange`:

https://github.com/dotnet/runtime/blob/526d72325979a59388d808f02672632aff30fe4e/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs#L251